### ISSUE TITLE
[Feature] Considering inheritance relation between actions.

### ISSUE THREAD

**Issue by khaes-kth** [SpoonLabs/coming#180]
Hi,

Currently, a change is an instance of a pattern only if the mentioned entity type in the pattern specification is exactly the same as the diff entity type. See [HERE](https://github.com/SpoonLabs/coming/blob/master/src/main/java/fr/inria/coming/changeminer/analyzer/instancedetector/DetectorChangePatternInstanceEngine.java#L461). However, it would be interesting if we could define a pattern by mentioning an entity type and retrieve all changes with that type of entity or any type of entity that extends the mentioned type.

For example, [this change](https://github.com/SpoonLabs/astor-experiments/tree/master/cardumen-patches/unformatted-patches/Math/Math28#patch-diff-3) includes an _invocation_ update (`minRatioPositions.add->Precision.equals`). However, it is also an instance of _expression_ update because [invocation extends expression](https://github.com/INRIA/spoon/blob/master/src/main/java/spoon/reflect/code/CtInvocation.java#L33).
It would be interesting to define a pattern as follows and detect all instances of expression updates (including invocation updates such as the mentioned one):

```
<pattern name="expression_replacement">
    <entity id="1" type="Expression" />
    <action entityId="1" type="UPD"/>
</pattern>
```

Currently, this pattern specification does not recognize [this change](https://github.com/SpoonLabs/astor-experiments/tree/master/cardumen-patches/unformatted-patches/Math/Math28#patch-diff-3) as a detected instance.

**Comment by martinezmatias** [comment]
Hi @khaes-kth 

> it would be interesting if we could define a pattern by mentioning an entity type and retrieve all changes with that type of entity or any type of entity that extends the mentioned type.

Yes, I completely agree that we need that functionality. 
One simple option that came to my mind is to create that hierarchy statically (in a Json File and load each time the tool starts, or even in the hardcoded). So that structure could be, for each element, the list of parent in the hierarchy.

**Comment by khaes-kth** [comment]
Yes, this is an option. But what about the following _dynamic_ option?
1- Use reflection to find the Gumtree class (such as CtExpression) corresponding to the pattern entity type.
2- Check if the found class is an `instanceof` the type of the diff entity. [Here](https://github.com/SpoonLabs/coming/blob/master/src/main/java/fr/inria/coming/changeminer/analyzer/instancedetector/DetectorChangePatternInstanceEngine.java#L461), `parentEntity.getEntityType()` gives the pattern entity type and [currentNodeFromAction](https://github.com/SpoonLabs/coming/blob/master/src/main/java/fr/inria/coming/changeminer/analyzer/instancedetector/DetectorChangePatternInstanceEngine.java#L439) is the diff entity.

**Comment by monperrus** [comment]
@khaes-kth will implement a first version

**Comment by khaes-kth** [comment]
This feature is added in #182

**Comment by martinezmatias** [comment]
Hi @khaes-kth 

I reopen the issue, until PR be merged.
BTW: which solution have you implemented? (The PR has many of commits for review)

**Comment by khaes-kth** [comment]
Hi @martinezmatias ,



> I reopen the issue, until PR be merged.

OK, Good.


> BTW: which solution have you implemented? (The PR has many of commits for review)

As you suggested, the static solution is implemented. The relations are stored [here](https://github.com/khaes-kth/coming/blob/master/src/main/resources/gumtree-inheritance-relations.txt) and used [here](https://github.com/khaes-kth/coming/blob/master/src/main/java/fr/inria/coming/changeminer/analyzer/instancedetector/DetectorChangePatternInstanceEngine.java#L463) and [here](https://github.com/khaes-kth/coming/blob/master/src/main/java/fr/inria/coming/utils/GumtreeHelper.java#L54).

**Comment by martinezmatias** [comment]
Hi @khaes-kth 
It seems perfect.
One minor refactor I propose is to rename the entity `GumtreeHelper `. IMHO, including Gumtree in the name is confusing because Gumtree is the algorithm to compute ast diff and this helper focuses on determining relations between the types of an AST.  
I would propose something similar to  `EntityTypeRelationshipResolver`.
WDYT?

**Comment by khaes-kth** [comment]
You are right. It is confusing. However, I had chosen that name because that class not only retrieved entity type relations, but it also had several other functions like [this](https://github.com/khaes-kth/coming/blob/6aa307f655012b868d3186d81dd2e79aafda3209/src/main/java/fr/inria/coming/utils/EntityTypesInfoResolver.java#L116) and [this](https://github.com/khaes-kth/coming/blob/6aa307f655012b868d3186d81dd2e79aafda3209/src/main/java/fr/inria/coming/repairability/JSONRepairabilityOutput.java#L128).
I just [renamed ](https://github.com/khaes-kth/coming/commit/6aa307f655012b868d3186d81dd2e79aafda3209)the `GumtreeHelper` class to `EntityTypesInfoResolver` and moved the `getOperationStats` function to the class where it is used.
Do you think it should have been refactored in a different way?

**Comment by martinezmatias** [comment]
Hi @khaes-kth 

> Do you think it should have been refactored in a different way?

It seems perfect!
Thanks a lot.
Matias
